src/java_bytecode/ci_lazy_methods.h

```diff
@@ -1,49 +1,107 @@
+  lazy_methodst;
 
-class ci_lazy_methodst
+class ci_lazy_methodst:public messaget
 {
 public:
   ci_lazy_methodst(
-    std::vector<irep_idt> &_needed_methods,
-    std::set<irep_idt> &_needed_classes,
-    symbol_tablet &_symbol_table):
-  needed_methods(_needed_methods),
-  needed_classes(_needed_classes),
-  symbol_table(_symbol_table)
-  {}
-
-  void add_needed_method(const irep_idt &);
-  // Returns true if new
-  bool add_needed_class(const irep_idt &);
+    const symbol_tablet &symbol_table,
+    const irep_idt &main_class,
+    const std::vector<irep_idt> &main_jar_classes,
+    const std::vector<irep_idt> &lazy_methods_extra_entry_points,
+    java_class_loadert &java_class_loader,
+    message_handlert &message_handler);
+
+  // not const since messaget
+  bool operator()(
+    symbol_tablet &symbol_table,
+    lazy_methodst &lazy_methods,
+    std::function<void(
+      const symbolt &,
+      const java_bytecode_parse_treet::methodt &,
+      ci_lazy_methods_neededt)> method_converter);
 
 private:
-  // needed_methods is a vector because it's used as a work-list
-  // which is periodically cleared. It can't be relied upon to
-  // contain all methods that have previously been elaborated.
-  // It should be changed to a set if we develop the need to use
-  // it that way.
-  std::vector<irep_idt> &needed_methods;
-  // needed_classes on the other hand is a true set of every class
-  // found so far, so we can use a membership test to avoid
-  // repeatedly exploring a class hierarchy.
-  std::set<irep_idt> &needed_classes;
-  symbol_tablet &symbol_table;
+  void resolve_method_names(
+    std::vector<irep_idt> &methods,
+    const symbol_tablet &symbol_table);
+
+  void initialize_needed_classes(
+    const std::vector<irep_idt> &entry_points,
+    const namespacet &ns,
+    ci_lazy_methods_neededt &lazy_methods);
+
+  void initialize_needed_classes_from_pointer(
+    const pointer_typet &pointer_type,
+    const namespacet &ns,
+    ci_lazy_methods_neededt &lazy_methods);
+
+  void gather_virtual_callsites(
+    const exprt &e,
+    std::vector<const code_function_callt *> &result);
+
+  void get_virtual_method_targets(
+    const code_function_callt &c,
+    const std::set<irep_idt> &needed_classes,
+    std::vector<irep_idt> &needed_methods,
+    symbol_tablet &symbol_table);
+
+  void gather_needed_globals(
+    const exprt &e,
+    const symbol_tablet &symbol_table,
+    symbol_tablet &needed);
+
+  void gather_field_types(const typet &class_type,
+    const namespacet &ns,
+    ci_lazy_methods_neededt &lazy_methods);
+
+  irep_idt get_virtual_method_target(
+    const std::set<irep_idt> &needed_classes,
+    const irep_idt &call_basename,
+    const irep_idt &classname,
+    const symbol_tablet &symbol_table);
+
+  static irep_idt build_virtual_method_name(
+    const irep_idt &class_name,
+    const irep_idt &component_method_name);
+
+  class_hierarchyt class_hierarchy;
+  const irep_idt main_class;
```

Comment:
If it's an STL-compatible functor it *must* be copyable. If not, it might be helpful to explicitly disable the copy operations, and to supply an explanatory comment.